torch/distributed/_tensor/random.py

```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+
+    .. warning::
+        When calling this function, :func:`manual_seed` must be called from all ranks of the
+        default `ProcessGroup` even if some ranks may not be a part of the `device_mesh`,
+        with the same `seed` value.
+        If ``device_mesh`` is a sub-mesh and the calling rank is not a part of it,
+        `manual_seed` will not set its GPU device's generator seed.
+        Current implementation only supports a GPU device mesh.
+    """
+    # allgather the seed from rank 0 over the default PG
+    object_list = [seed] * dist.get_world_size()
+    dist.all_gather_object(object_list, seed)
+    for rank, object in enumerate(object_list):
+        if seed != int(object):
+            raise RuntimeError(
+                f"calling manual_seed function over {device_mesh} but received different seed values on ranks:",
+                f"seed on rank {dist.get_rank()} is {seed}, and seed on rank {rank} is {object}!",
+            )
+
+    # the current rank is in mesh
+    if device_mesh.get_coordinate() is not None:
+        if isinstance(_parallel_rng_style, MegatronRNG):
+            _parallel_rng_style._manual_seed(device_mesh, seed, tp_dim)
+        elif isinstance(_parallel_rng_style, DTensorRNG):
+            _parallel_rng_style._manual_seed(seed)
+        else:
+            raise RuntimeError("Unknown parallel RNG style!")
+
+
+# TODO: change this to ABC
+class ParallelRNGStyle:
+    def __init__(self):
+        pass
+
+    # override this function to define custom parallel random number generation
+    # to execute DTensor computation
+    @contextlib.contextmanager
+    def _parallel_region(self):
+        pass
+
+class DTensorRNG(ParallelRNGStyle):
+    # The default parallel random number generation style
+    def __init__(self):
+        super().__init__()
+
+    def _manual_seed(self, parallel_seed: int) -> None:
+        # do not pollute torch's default RNG state
+        device = torch.cuda.current_device()
+        with torch.random.fork_rng([device]):
+            torch.cuda.manual_seed(parallel_seed)
+            _rng_tracker.set_rng_state("parallel-rng", torch.cuda.get_rng_state())
+
+    @contextlib.contextmanager
+    def _parallel_region(self, spec: DTensorSpec):
+        # check if the parallel rng state has been synchronized or not
+        if not self.rng_state_is_sync("parallel-rng"):
+            raise RuntimeError(
+                "DTensorRNG requires random seed to be synchronized"
+                "before entering into a parallel region!"
+            )
+
+        # TODO: rewrite this set/restore logic using fork_rng()?
+        # fork rng state first since parallel RNG state should be used instead
+        old_state = torch.cuda.get_rng_state()  # a 16-byte ByteTensor
+        torch.cuda.set_rng_state(self.rng_states["parallel-rng"])
+        # set offset before entering the parallel region
+        mesh = spec.mesh
+        old_offset = _get_rng_offset(mesh)
+        set_pre_op_offset(spec)
```

Comment:
I think we should add all those methods in this file to this class, and rename this to `DTensorRNGTracker` instead.

Reply:
I can move them into DtensorRNG. But merging tracker and parallel style caused circular dependency.